<launch>

  <!-- ENV VARS -->
  <arg name="UAV_NAME" default="$(env UAV_NAME)" />
  <arg name="enable_profiler" default="$(optenv PROFILER false)" />

  <!-- DEFAULT PARAMETERS AND FRAMES -->
  <arg name="fcu_frame" default="$(arg UAV_NAME)/fcu"/>
  <arg name="lidar_frame" default="$(arg UAV_NAME)/os1_lidar"/>
  <arg name="odom_frame" default="$(arg UAV_NAME)/aloam_odom_origin"/>
  <arg name="map_frame" default="$(arg UAV_NAME)/aloam_origin"/>

  <!-- SENSOR PARAMETERS -->
  <arg name="points_topic" default="os1_cloud_node/points_processed" />
  <arg name="number_of_rings" default="16" />
  <arg name="vertical_fov" default="33.2" />
  <arg name="sensor_frequency" default="10.0" />

  <!-- MAPPING PARAMETERS -->
  <arg name="mapping_rate" default="$(arg sensor_frequency)" />
  <arg name="map_publish_rate" default="0.5" />
  <arg name="mapping_line_resolution" default="0.4" />
  <arg name="mapping_plane_resolution" default="0.8" />

  <!-- LAUNCH PARAMETERS -->
  <arg name="verbose" default="false" />
  <arg name="nodelet" default="standalone" />
  <arg name="nodelet_manager" default="" />
  <arg name="launch_delay" default=""/>
  <arg name="launch_prefix" default=""/>

  <group ns="$(arg UAV_NAME)">
    
    <node pkg="nodelet" type="nodelet" name="aloam_slam" args="$(arg nodelet) aloam_slam/AloamSlam $(arg nodelet_manager)" output="screen" launch-prefix="bash -c 'sleep $(arg launch_delay); $0 $@'; $(arg launch_prefix)">
      <param name="verbose" type="bool" value="$(arg verbose)"/>
      <param name="enable_profiler" type="bool" value="$(arg enable_profiler)" />

      <param name="scan_line" type="int" value="$(arg number_of_rings)" />
      <param name="vertical_fov" type="double" value="$(arg vertical_fov)"/>
      <param name="sensor_frequency" type="double" value="$(arg sensor_frequency)"/>

      <param name="uav_name" type="string" value="$(arg UAV_NAME)"/>
      <param name="fcu_frame" type="string" value="$(arg fcu_frame)"/>
      <param name="odom_frame" type="string" value="$(arg odom_frame)"/>
      <param name="map_frame" type="string" value="$(arg map_frame)"/>
      <param name="lidar_frame" type="string" value="$(arg lidar_frame)"/>

      <param name="mapping_rate" type="double" value="$(arg mapping_rate)" />
      <param name="map_publish_rate" type="double" value="$(arg map_publish_rate)" />
      <!-- <param name="mapping_line_resolution" type="double" value="$(arg mapping_line_resolution)" /> -->
      <!-- <param name="mapping_plane_resolution" type="double" value="$(arg mapping_plane_resolution)" /> -->
      <param name="mapping_resolution/corners/min" type="double" value="0.4" />
      <param name="mapping_resolution/corners/max" type="double" value="0.4" />
      <param name="mapping_resolution/surfs/min" type="double" value="0.8" />
      <param name="mapping_resolution/surfs/max" type="double" value="0.8" />
      
      <param name="feature_selection/enable" type="bool" value="true" />
      <param name="feature_selection/local_segments" type="int" value="1" />
      <param name="feature_selection/min_count_percent" type="double" value="0.1" />
      <param name="feature_selection/corners/keep_percentile" type="double" value="0.1" />
      <param name="feature_selection/corners/keep_standalone" type="bool" value="false" />
      <param name="feature_selection/surfs/keep_percentile" type="double" value="0.1" />
      <param name="feature_selection/surfs/keep_standalone" type="bool" value="true" />
      <!-- <param name="feature_selection/surfs/gradient/limit/upper" type="double" value="1.0" /> -->
      <!-- <param name="feature_selection/surfs/gradient/limit/bottom" type="double" value="0.4" /> -->
      <!-- <param name="feature_selection/corners/gradient/limit/upper" type="double" value="1.0" /> -->
      <!-- <param name="feature_selection/corners/gradient/limit/bottom" type="double" value="0.05" /> -->

      <!-- Subscribers -->
      <remap from="~laser_cloud_in" to="$(arg points_topic)"/>
      <remap from="~orientation_in" to="odometry/orientation"/>

      <!-- Publishers -->
      <remap from="~odom_local_out" to="aloam/odom_local"/>
      <remap from="~path_out" to="aloam/path"/>
      <remap from="~odom_global_out" to="aloam/odom"/>
      <remap from="~map_out" to="aloam/map"/>
      <remap from="~diag_out" to="aloam/diag"/>
      <remap from="~scan_registered_out" to="aloam/scan_registered"/>
      <remap from="~features_corners_sharp_out" to="aloam/features/corners/sharp"/>
      <remap from="~features_corners_less_sharp_out" to="aloam/features/corners/sharpless"/>
      <remap from="~features_surfs_flat_out" to="aloam/features/surfs/flat"/>
      <remap from="~features_surfs_less_flat_out" to="aloam/features/surfs/flatless"/>
      <remap from="~features_corners_selected_out" to="aloam/features/corners/selected"/>
      <remap from="~features_surfs_selected_out" to="aloam/features/surfs/selected"/>
        
      <!-- Service servers -->
      <remap from="~srv_reset_mapping_in" to="aloam/reset/mapping"/>

    </node>

  </group>

</launch>
